piptools/build.py

```diff
@@ -0,0 +1,159 @@
+                builder=builder,
+                distributions=build_distributions,
+                package_name=_get_name(metadata),
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: str, *, isolated: bool, quiet: bool
+) -> Iterator[build.ProjectBuilder]:
+    builder = build.ProjectBuilder(
+        os.fspath(src_dir),
+        runner=pyproject_hooks.quiet_subprocess_runner
+        if quiet
+        else pyproject_hooks.default_subprocess_runner,
+    )
+
+    if not isolated:
+        yield builder
+        return
+
+    with build.env.IsolatedEnvBuilder() as env:
+        builder.python_executable = env.executable
+        builder.scripts_dir = env.scripts_dir
+        env.install(builder.build_system_requires)
+        env.install(builder.get_requires_for_build("wheel"))
+        yield builder
+
+
+def _build_project_wheel_metadata(
+    builder: build.ProjectBuilder,
+) -> Any:
+    with tempfile.TemporaryDirectory() as tmpdir:
+        path = pathlib.Path(builder.metadata_path(tmpdir))
+        return importlib_metadata.PathDistribution(path).metadata
+
+
+def _get_name(metadata: Message) -> str:
+    return metadata.get_all("Name")[0]  # type: ignore
+
+
+def _prepare_requirements(
+    metadata: Message, src_file: str
+) -> Iterator[InstallRequirement]:
+    package_name = _get_name(metadata)
+    src_name = os.path.basename(src_file)
+    comes_from = f"{package_name} ({src_name})"
+
+    for req in metadata.get_all("Requires-Dist") or []:
+        parts = parse_req_from_line(req, comes_from)
+        if parts.requirement.name == package_name:
+            package_dir = os.path.dirname(os.path.abspath(src_file))
+            # Replace package name with package directory in the requirement
+            # string so that pip can find the package as self-referential.
+            # Note the string can contain extras, so we need to replace only
+            # the package name, not the whole string.
+            replaced_package_name = req.replace(package_name, package_dir, 1)
+            parts = parse_req_from_line(replaced_package_name, comes_from)
+
+        yield InstallRequirement(
+            parts.requirement,
+            comes_from,
+            link=parts.link,
+            markers=parts.markers,
+            extras=parts.extras,
+        )
+
+
+def _prepare_build_requirements(
+    builder: build.ProjectBuilder,
+    distributions: tuple[str, ...],
+    package_name: str,
+) -> Iterator[InstallRequirement]:
+    result = collections.defaultdict(set)
+
+    # Build requirements will only be present if a pyproject.toml file exists,
+    # but if there is also a setup.py file then only that will be explicitly
+    # processed due to the order of `DEFAULT_REQUIREMENTS_FILES`.
+    src_name = PYPROJECT_TOML
```

Comment:
If we use only the file name as `src_file`, we lose the relative path. I'd expect to see the relative path to the project file in the annotations, the same as in other cases:

```commandline
$ pip-compile tmp/requirements.in --no-header
six==1.16.0
    # via -r tmp/requirements.in
    
$ pip-compile tmp/pyproject.toml --no-header
six==1.16.0
    # via tmp (tmp/pyproject.toml)
```

Reply:
See original implementation: https://github.com/apljungquist/pip-tools/pull/1/files#diff-eff8258d1e71cf62b03337ca654093501cb23d92dd9c2894091a6175850a660dR136-R140

Reply:
Hmm, I removed it because the absolute path was showing up in the output but now I cannot reproduce that. Will fix.

Reply:
Thanks!